### PR TITLE
TEPHRA-105: Adding HBase 1.1 compatibility module

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Adds supports for HBase1.1
- Tested on HBase1.1 cluster using smoke test from here: https://github.com/caskdata/tephra/pull/75

Issue: https://issues.cask.co/browse/TEPHRA-105

### CODE REVIEW

File: tephra-hbase-compat-1.1/pom.xml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    <groupId>co.cask.tephra</groupId>
+    <version>0.6.2-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>tephra-hbase-compat-1.1</artifactId>
+  <name>Tephra HBase 1.1 Compatibility</name>
+
+  <properties>
+    <hadoop.version>2.4.0</hadoop.version>
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/pom.xml]: Why is this needed?

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      hBaseAdmin = new HBaseAdmin(hTable.getConfiguration());
+      if (!hBaseAdmin.tableExists(secondaryIndexTableName)) {
+        hBaseAdmin.createTable(new HTableDescriptor(secondaryIndexTableName));
+      }
+      secondaryIndexHTable = new HTable(hTable.getConfiguration(), secondaryIndexTableName);
+    } catch (Exception e) {
+      Throwables.propagate(e);
+    } finally {
+      try {
+        hBaseAdmin.close();
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java]: Use the Java 7 ARM syntax.

```java
try (HBaseAdmin admin = new HBaseAdmin(...)) {
}
```

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A Transactional SecondaryIndexTable.
+ */
+public class SecondaryIndexTable {
+  private byte[] secondaryIndex;
+  private TransactionAwareHTable transactionAwareHTable;
+  private TransactionAwareHTable secondaryIndexTable;
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java]: Who is responsible in closing this table? This class doesn't implements Closeable.

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A Transactional SecondaryIndexTable.
+ */
+public class SecondaryIndexTable {
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java]: This index table implementation is not good. Shouldn't it implements TransactionAware? I know this class is copied from existing compat modules, which suggest all existing implementations are not good. Maybe open a JIRA have that fixed.

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/TransactionAwareHTable.java [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+
+  @Override
+  protected boolean doRollback() throws Exception {
+    try {
+      // pre-size arraylist of deletes
+      int size = 0;
+      for (Set<ActionChange> cs : changeSets.values()) {
+        size += cs.size();
+      }
+      List<Delete> rollbackDeletes = new ArrayList<Delete>(size);
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/TransactionAwareHTable.java]: Just `new ArrayList<>(size);`. You should be seeing warning in IntelliJ (yellow). Please pay attention to warnings.

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/TransactionAwareHTable.java [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+    }
+    hTable.delete(transactionalizeAction(delete));
+  }
+
+  @Override
+  public void delete(List<Delete> deletes) throws IOException {
+    if (tx == null) {
+      throw new IOException("Transaction not started");
+    }
+    List<Delete> transactionalizedDeletes = new ArrayList<Delete>(deletes.size());
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/TransactionAwareHTable.java]: Just `new ArrayList<>`.

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionProcessor.java [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+ * codec.addToOperation(get, t);
+ * }
+ * </p>
+ */
+public class TransactionProcessor extends BaseRegionObserver {
+  private static final Log LOG = LogFactory.getLog(TransactionProcessor.class);
+
+  private TransactionStateCache cache;
+  private final TransactionCodec txCodec;
+  protected Map<byte[], Long> ttlByFamily = Maps.newTreeMap(Bytes.BYTES_COMPARATOR);
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionProcessor.java]: final. Also, why protected instead of private?

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionVisibilityFilter.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  @Override
+  public void reset() {
+    deleteTracker.reset();
+  }
+
+  @Override
+  public byte[] toByteArray() throws IOException {
+    return super.toByteArray();
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionVisibilityFilter.java]: Why need to override if just calling super?

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java (hunk 4 above)
Comment (rohitsinha54) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java]: Jira: https://issues.cask.co/browse/TEPHRA-124

File: tephra-hbase-compat-1.1/pom.xml (hunk 1 above)
Comment (rohitsinha54) [inline tephra-hbase-compat-1.1/pom.xml]: HBase1.1 is only tested on 2.4 +
Added a comment like 1.0
http://hbase.apache.org/book.html#hadoop

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+      } catch (TransactionFailureException e1) {
+        throw new IOException("Could not rollback transaction", e1);
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    transactionAwareHTable.close();
+    secondaryIndexTable.close();
```
Comment (chtyim) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/SecondaryIndexTable.java]: You should catch the exception for the first close to make sure second close always get executed.

```java
try {
  first.close();
} catch (IOException e) {
  try {
    second.close();
  } catch (IOException ex) {
    e.addSuppressed(e);
  }
  throw e;
}
second.close();
```

**Comment by chtyim** [comment]
Just one comment left. Rest LGTM.

**Comment by rohitsinha54** [comment]
Fixed all comments. 
Testing one last time with smoke test and in CDAP before merging this.

File: tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionProcessor.java (hunk 7 above)
Comment (rohitsinha54) [inline tephra-hbase-compat-1.1/src/main/java/co/cask/tephra/hbase11/coprocessor/TransactionProcessor.java]: We discussed this in person. They are used in CDAP
